In [32]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import precision_score,recall_score,f1_score,accuracy_score,confusion_matrix

In [33]:
df=pd.read_csv("loan_data.csv")

In [34]:
df.head()

person_age person_gender person_education  person_income  person_emp_exp  \
0        22.0        female           Master        71948.0               0   
1        21.0        female      High School        12282.0               0   
2        25.0        female      High School        12438.0               3   
3        23.0        female         Bachelor        79753.0               0   
4        24.0          male           Master        66135.0               1   

  person_home_ownership  loan_amnt loan_intent  loan_int_rate  \
0                  RENT    35000.0    PERSONAL          16.02   
1                   OWN     1000.0   EDUCATION          11.14   
2              MORTGAGE     5500.0     MEDICAL          12.87   
3                  RENT    35000.0     MEDICAL          15.23   
4                  RENT    35000.0     MEDICAL          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.49                         3.0           561   
1                 0.08                         2.0           504   
2                 0.44                         3.0           635   
3                 0.44                         2.0           675   
4                 0.53                         4.0           586   

  previous_loan_defaults_on_file  loan_status  
0                             No            1  
1                            Yes            0  
2                             No            1  
3                             No            1  
4                             No            1

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45000 entries, 0 to 44999
Data columns (total 14 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   person_age                      45000 non-null  float64
 1   person_gender                   45000 non-null  object 
 2   person_education                45000 non-null  object 
 3   person_income                   45000 non-null  float64
 4   person_emp_exp                  45000 non-null  int64  
 5   person_home_ownership           45000 non-null  object 
 6   loan_amnt                       45000 non-null  float64
 7   loan_intent                     45000 non-null  object 
 8   loan_int_rate                   45000 non-null  float64
 9   loan_percent_income             45000 non-null  float64
 10  cb_person_cred_hist_length      45000 non-null  float64
 11  credit_score                    45000 non-null  int64  
 12  previous_loan_defaults_on_file  

In [36]:
df.isnull().sum()

person_age                        0
person_gender                     0
person_education                  0
person_income                     0
person_emp_exp                    0
person_home_ownership             0
loan_amnt                         0
loan_intent                       0
loan_int_rate                     0
loan_percent_income               0
cb_person_cred_hist_length        0
credit_score                      0
previous_loan_defaults_on_file    0
loan_status                       0
dtype: int64

In [37]:
x=df.drop('loan_status',axis=1)
y=df['loan_status']

In [38]:
xtrain,xtest,ytrain,ytest=train_test_split(x,y,test_size=0.2,random_state=42)

In [39]:
num_cols= xtrain.select_dtypes(exclude="object").columns
cat_cols=xtrain.select_dtypes(include="object").columns

In [40]:
scaler=StandardScaler()
encode=OneHotEncoder()

In [41]:
xtrain[num_cols]=scaler.fit_transform(xtrain[num_cols])
xtest[num_cols]=scaler.transform(xtest[num_cols])

In [42]:
xtrain

person_age person_gender person_education  person_income  \
25180    1.039475        female         Bachelor       0.194225   
12555   -0.456186          male      High School      -0.089962   
29153    2.202767        female           Master       2.828167   
23838   -0.123817          male        Associate       0.158772   
35686   -0.123817          male           Master       0.049373   
...           ...           ...              ...            ...   
11284   -0.290001          male      High School       0.091197   
44732   -0.456186        female      High School      -0.536266   
38158    0.873291        female        Associate      -0.261045   
860     -0.290001          male           Master       1.144986   
15795   -0.290001        female      High School       1.522453   

       person_emp_exp person_home_ownership  loan_amnt        loan_intent  \
25180        0.931186              MORTGAGE   0.859025           PERSONAL   
12555       -0.395014                  RENT   0.383393            VENTURE   
29153        2.091610              MORTGAGE   2.285919           PERSONAL   
23838       -0.229239                  RENT   0.002888          EDUCATION   
35686        0.268086                  RENT  -0.399971    HOMEIMPROVEMENT   
...               ...                   ...        ...                ...   
11284        0.102311              MORTGAGE   0.165396          EDUCATION   
44732       -0.395014              MORTGAGE  -0.587529    HOMEIMPROVEMENT   
38158        0.599636              MORTGAGE   0.178396            VENTURE   
860          0.102311                  RENT   1.651744  DEBTCONSOLIDATION   
15795       -0.229239              MORTGAGE   0.859025           PERSONAL   

       loan_int_rate  loan_percent_income  cb_person_cred_hist_length  \
25180       0.577910             0.119054                    0.816732   
12555       0.286925             0.233921                   -0.478914   
29153      -0.211430            -0.799885                    1.334990   
23838       2.052907            -0.455283                    0.298474   
35686       0.658182            -0.685018                   -0.738043   
...              ...                  ...                         ...   
11284      -1.663014            -0.225548                   -0.478914   
44732       0.230066             0.348788                   -0.478914   
38158      -0.087677             0.463655                    0.816732   
860         2.337203            -0.340416                   -0.738043   
15795       0.002628            -0.799885                   -0.738043   

       credit_score previous_loan_defaults_on_file  
25180     -0.032444                             No  
12555      0.524687                            Yes  
29153      1.002227                            Yes  
23838      0.166531                             No  
35686      1.459871                            Yes  
...             ...                            ...  
11284     -1.465066                            Yes  
44732      0.285916                             No  
38158      0.385404                            Yes  
860       -0.569677                             No  
15795      0.624174                            Yes  

[36000 rows x 13 columns]

In [43]:
encoded_xtrain=encode.fit_transform(xtrain[cat_cols])
xtrain[encode.get_feature_names_out()]=encoded_xtrain.toarray()

In [44]:
encode.get_feature_names_out()

array(['person_gender_female', 'person_gender_male',
       'person_education_Associate', 'person_education_Bachelor',
       'person_education_Doctorate', 'person_education_High School',
       'person_education_Master', 'person_home_ownership_MORTGAGE',
       'person_home_ownership_OTHER', 'person_home_ownership_OWN',
       'person_home_ownership_RENT', 'loan_intent_DEBTCONSOLIDATION',
       'loan_intent_EDUCATION', 'loan_intent_HOMEIMPROVEMENT',
       'loan_intent_MEDICAL', 'loan_intent_PERSONAL',
       'loan_intent_VENTURE', 'previous_loan_defaults_on_file_No',
       'previous_loan_defaults_on_file_Yes'], dtype=object)

In [45]:
encoded_xtest=encode.transform(xtest[cat_cols])
xtest[encode.get_feature_names_out()]=encoded_xtest.toarray()

In [46]:
xtrain = xtrain.drop(columns=cat_cols)
xtest = xtest.drop(columns=cat_cols)

In [47]:
xtrain_cat_df=pd.DataFrame(encoded_xtrain.toarray(),columns=encode.get_feature_names_out())
xtest_cat_df=pd.DataFrame(encoded_xtest.toarray(),columns=encode.get_feature_names_out())

In [48]:
xtrain_cat_df

person_gender_female  person_gender_male  person_education_Associate  \
0                       1.0                 0.0                         0.0   
1                       0.0                 1.0                         0.0   
2                       1.0                 0.0                         0.0   
3                       0.0                 1.0                         1.0   
4                       0.0                 1.0                         0.0   
...                     ...                 ...                         ...   
35995                   0.0                 1.0                         0.0   
35996                   1.0                 0.0                         0.0   
35997                   1.0                 0.0                         1.0   
35998                   0.0                 1.0                         0.0   
35999                   1.0                 0.0                         0.0   

       person_education_Bachelor  person_education_Doctorate  \
0                            1.0                         0.0   
1                            0.0                         0.0   
2                            0.0                         0.0   
3                            0.0                         0.0   
4                            0.0                         0.0   
...                          ...                         ...   
35995                        0.0                         0.0   
35996                        0.0                         0.0   
35997                        0.0                         0.0   
35998                        0.0                         0.0   
35999                        0.0                         0.0   

       person_education_High School  person_education_Master  \
0                               0.0                      0.0   
1                               1.0                      0.0   
2                               0.0                      1.0   
3                               0.0                      0.0   
4                               0.0                      1.0   
...                             ...                      ...   
35995                           1.0                      0.0   
35996                           1.0                      0.0   
35997                           0.0                      0.0   
35998                           0.0                      1.0   
35999                           1.0                      0.0   

       person_home_ownership_MORTGAGE  person_home_ownership_OTHER  \
0                                 1.0                          0.0   
1                                 0.0                          0.0   
2                                 1.0                          0.0   
3                                 0.0                          0.0   
4                                 0.0                          0.0   
...                               ...                          ...   
35995                             1.0                          0.0   
35996                             1.0                          0.0   
35997                             1.0                          0.0   
35998                             0.0                          0.0   
35999                             1.0                          0.0   

       person_home_ownership_OWN  person_home_ownership_RENT  \
0                            0.0                         0.0   
1                            0.0                         1.0   
2                            0.0                         0.0   
3                            0.0                         1.0   
4                            0.0                         1.0   
...                          ...                         ...   
35995                        0.0                         0.0   
35996                        0.0                         0.0   
35997                        0.0                         0.0   
35998                        0.0       

In [49]:
xtrain_concat=pd.concat([xtrain.reset_index(drop=True),xtrain_cat_df.reset_index(drop=True)],axis=1)
xtest_concat=pd.concat([xtest.reset_index(drop=True),xtest_cat_df.reset_index(drop=True)],axis=1)

In [50]:
xtrain_concat

person_age  person_income  person_emp_exp  loan_amnt  loan_int_rate  \
0        1.039475       0.194225        0.931186   0.859025       0.577910   
1       -0.456186      -0.089962       -0.395014   0.383393       0.286925   
2        2.202767       2.828167        2.091610   2.285919      -0.211430   
3       -0.123817       0.158772       -0.229239   0.002888       2.052907   
4       -0.123817       0.049373        0.268086  -0.399971       0.658182   
...           ...            ...             ...        ...            ...   
35995   -0.290001       0.091197        0.102311   0.165396      -1.663014   
35996   -0.456186      -0.536266       -0.395014  -0.587529       0.230066   
35997    0.873291      -0.261045        0.599636   0.178396      -0.087677   
35998   -0.290001       1.144986        0.102311   1.651744       2.337203   
35999   -0.290001       1.522453       -0.229239   0.859025       0.002628   

       loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.119054                    0.816732     -0.032444   
1                 0.233921                   -0.478914      0.524687   
2                -0.799885                    1.334990      1.002227   
3                -0.455283                    0.298474      0.166531   
4                -0.685018                   -0.738043      1.459871   
...                    ...                         ...           ...   
35995            -0.225548                   -0.478914     -1.465066   
35996             0.348788                   -0.478914      0.285916   
35997             0.463655                    0.816732      0.385404   
35998            -0.340416                   -0.738043     -0.569677   
35999            -0.799885                   -0.738043      0.624174   

       person_gender_female  person_gender_male  ...  \
0                       1.0                 0.0  ...   
1                       0.0                 1.0  ...   
2                       1.0                 0.0  ...   
3                       0.0                 1.0  ...   
4                       0.0                 1.0  ...   
...                     ...                 ...  ...   
35995                   0.0                 1.0  ...   
35996                   1.0                 0.0  ...   
35997                   1.0                 0.0  ...   
35998                   0.0                 1.0  ...   
35999                   1.0                 0.0  ...   

       person_home_ownership_OWN  person_home_ownership_RENT  \
0                            0.0                         0.0   
1                            0.0                         1.0   
2                            0.0                         0.0   
3                            0.0                         1.0   
4                            0.0                         1.0   
...                          ...                         ...   
35995                        0.0                         0.0   
35996                        0.0                         0.0   
35997                        0.0                         0.0   
35998                        0.0                         1.0   
35999                        0.0                         0.0   

       loan_intent_DEBTCONSOLIDATION  loan_intent_EDUCATION  \
0                                0.0                    0.0   
1                                0.0                    0.0   
2                                0.0                    0.0   
3                                0.0                    1.0   
4                                0.0                    0.0   
...                              ...                    ...   
35995                            0.0                    1.0   
35996                            0.0                    0.0   
35997                            0.0                    0.0   
35998                            1.0                    0.0   
35999                            0.0                    0.0   

      

In [51]:
xtest_concat

person_age  person_income  person_emp_exp  loan_amnt  loan_int_rate  \
0       0.707106       0.189549        0.765411  -0.330054      -1.659669   
1      -0.622370      -0.278333        0.102311  -0.092238       0.163172   
2      -0.954740      -0.506739       -0.892339  -0.726413       1.297013   
3      -0.788555      -0.475810       -0.726564  -0.726413       0.999338   
4       2.368952      -0.148162        2.423160   0.066306       1.474280   
...          ...            ...             ...        ...            ...   
8995   -1.120924       0.614333       -0.892339   0.700481      -0.338527   
8996    1.870398      -0.219291        1.925835  -0.452925       1.822125   
8997   -0.622370      -0.019862       -0.726564   0.700481       0.216687   
8998    1.704214      -0.508563        1.760060  -0.250782       0.002628   
8999   -0.456186      -0.531076       -0.892339  -0.171510       0.046109   

      loan_percent_income  cb_person_cred_hist_length  credit_score  \
0               -0.685018                    1.075861     -0.629370   
1                0.233921                   -0.478914      0.285916   
2               -0.110681                   -0.478914      1.559358   
3               -0.110681                   -0.738043     -0.708960   
4                0.119054                    1.334990     -0.927833   
...                   ...                         ...           ...   
8995            -0.340416                   -0.997172      0.425199   
8996            -0.340416                    1.334990     -0.529882   
8997             0.463655                   -0.738043     -0.370702   
8998             0.923125                    2.112377      1.240998   
8999             1.152859                   -0.738043      0.047146   

      person_gender_female  person_gender_male  ...  \
0                      0.0                 1.0  ...   
1                      0.0                 1.0  ...   
2                      1.0                 0.0  ...   
3                      1.0                 0.0  ...   
4                      0.0                 1.0  ...   
...                    ...                 ...  ...   
8995                   0.0                 1.0  ...   
8996                   0.0                 1.0  ...   
8997                   1.0                 0.0  ...   
8998                   0.0                 1.0  ...   
8999                   1.0                 0.0  ...   

      person_home_ownership_OWN  person_home_ownership_RENT  \
0                           0.0                         0.0   
1                           0.0                         1.0   
2                           0.0                         1.0   
3                           1.0                         0.0   
4                           0.0                         0.0   
...                         ...                         ...   
8995                        0.0                         0.0   
8996                        0.0                         1.0   
8997                        0.0                         1.0   
8998                        0.0                         1.0   
8999                        0.0                         1.0   

      loan_intent_DEBTCONSOLIDATION  loan_intent_EDUCATION  \
0                               0.0                    1.0   
1                               0.0                    1.0   
2                               0.0                    0.0   
3                               0.0                    0.0   
4                               0.0                    0.0   
...                             ...                    ...   
8995                            0.0                    0.0   
8996                            0.0                    1.0   
8997                            0.0                    1.0   
8998                            0.0                    0.0   
8999                            0.0                    0.0   

      loan_intent_HOMEIMPROVEMENT  loan_intent_MEDICAL  loan_inte

In [52]:
model=LogisticRegression()

In [53]:
model.fit(xtrain_concat,ytrain)

LogisticRegression()

In [54]:
ypred=model.predict(xtest_concat)

In [55]:
model.score(xtest_concat,ytest)

0.8943333333333333

## Model evaluation

In [56]:
confusion_matrix(ytest,ypred)

array([[6558,  432],
       [ 519, 1491]], dtype=int64)

In [57]:
precision_score(ytest,ypred,pos_label=1)

0.7753510140405616

** Out of all the predictions approved loans (Class 1) was 0.775. Meaning 77% approved predictions were Truely approved.

In [58]:
precision_score(ytest,ypred,pos_label=0)

0.9266638406104282

** Out of all the predictions labeled as "Loan Not Approved", 92% were truly not approved.

In [59]:
recall_score(ytest,ypred,pos_label=1)

0.7417910447761195

** Out of all the actual "Loan Approved", 74% were Truely approved

In [60]:
recall_score(ytest,ypred,pos_label=0)

0.9381974248927039

**Out of all the actual "Loan Not Approved", 93% were Truely not approved

In [61]:
f1_score(ytest,ypred,pos_label=1) 

0.7581998474446987

** The F1-score for this class 1(Loan Approved) is 75%, a good balance between precision and recall

In [62]:
f1_score(ytest,ypred,pos_label=0) 

0.9323949669439113

** The F1-score for this class 0(Loan Not Approved) is 93%, showing strong balance between precision and recall

## Conclusion: The model give good result